src/Microsoft.AspNetCore.TestHost/ClientHandler.cs

```diff
@@ -187,27 +187,33 @@ internal void AbortRequest()
                 _responseStream.Complete();
             }
 
-            internal void CompleteResponse()
+            internal async Task CompleteResponseAsync()
             {
                 _pipelineFinished = true;
-                ReturnResponseMessage();
+                await ReturnResponseMessageAsync();
                 _responseStream.Complete();
-                _responseFeature.FireOnResponseCompleted();
+                await _responseFeature.FireOnResponseCompletedAsync();
             }
 
-            internal void ReturnResponseMessage()
+            internal async Task ReturnResponseMessageAsync()
             {
-                if (!_responseTcs.Task.IsCompleted)
+                // Check if the response has already started because the TrySetResult below could happen a bit late
+                // (as it happens on a different thread) by which point the CompleteResponseAsync could run and calls this
+                // method again.
+                if (!Context.HttpContext.Response.HasStarted)
                 {
-                    var response = GenerateResponse();
-                    // Dispatch, as TrySetResult will synchronously execute the waiters callback and block our Write.
-                    Task.Factory.StartNew(() => _responseTcs.TrySetResult(response));
+                    if (!_responseTcs.Task.IsCompleted)
```

Comment:
You shouldn't need to check both HasStarted and IsCompleted. HasStarted should be sufficient.